### DISCUSSION TITLE
SearXNG systemd

### DISCUSSION THREAD

**Discussion by nicfab** [searxng/searxng#1780] [Q&A]
Hello!

I changed the installation of my instance from docker to scripts.
The docker installation considers systemd.
I wondered whether I could set up a systemd file to allow starting SearXNG in case of a machine reboot.
I would appreciate any systemd template or sample content you can suggest.

**Comment by return42** [comment]
> I changed the installation of my instance from docker to scripts.

Hi @nicfab I moved this discussion to the searxng repo.

I'm not sure I understand you but if you look at the [architecture](https://docs.searxng.org/admin/architecture.html#uwsgi-setup) you will see that the scripts install SearXNG as a uWSGI app.

By default the scripts do install a [uWSGI setup](https://docs.searxng.org/admin/installation-uwsgi.html#uwsgi-setup) that listens on a socket. And the HTTP server connect to this socket:

- [NGINX’s SearXNG site](https://docs.searxng.org/admin/installation-nginx.html#nginx-s-searxng-site)
- [Apache’s SearXNG site](https://docs.searxng.org/admin/installation-apache.html#apache-s-searxng-site)

**Reply by nicfab** [reply]
Thank you.

I use Nginx and set the HTTP configuration instead of the socket.

My question was only related to whether it is necessary to set up a systemd config file or not to allow SearXNG to start automatically after a reboot.

**Reply by return42** [reply]
No its not needed, the uWSGI service is already started by your OS when you reboot. Services like SearXNG are *uWSGI-Apps*:

- read [Origin uWSGI](https://docs.searxng.org/admin/installation-uwsgi.html#origin-uwsgi) and the following sections to understand how uWSGI is implemented by your distributor,
- in [uWSGI maintenance](https://docs.searxng.org/admin/installation-uwsgi.html#uwsgi-maintenance) select your distribution and see how you can maintain a *uWSGI-App* (searxng).

**Reply by return42** [reply]
> I use Nginx and set the HTTP configuration instead of the socket.

Don't forget to set your searxng (*uWSGI-App*) listen on [`http = ..`](https://uwsgi-docs.readthedocs.io/en/latest/HTTP.html#the-uwsgi-http-https-router)

**Reply by nicfab** [reply]
> > I use Nginx and set the HTTP configuration instead of the socket.
> 
> Don't forget to set your searxng (_uWSGI-App_) listen on [`http = ..`](https://uwsgi-docs.readthedocs.io/en/latest/HTTP.html#the-uwsgi-http-https-router)

Can you please let me know what setting you are referring to?
Can you provide me with an example, please?

**Reply by return42** [reply]
> I use Nginx and set the HTTP configuration instead of the socket.

if this means, your HTTP service connects to the uWSGI driven SearXNG instance via TCP, then you need to set TCP in both configuration.

In this image the reverse proxy (your nginx) connects to the uWSGI-App SearXNG either with TCP socket (aka HTTP) or with unix socket

![image](https://user-images.githubusercontent.com/554536/190854573-4808ba2b-b007-499d-89f2-6d44e26f26b7.png)


## nginx / uWSGI-App SearXNG listens on a TCP socket (aka HTTP)

https://github.com/searxng/searxng/blob/b3708e4137cf66c7f2966e811c2fae907c28089a/utils/templates/etc/nginx/default.apps-available/searxng.conf#L3

https://github.com/searxng/searxng/blob/b3708e4137cf66c7f2966e811c2fae907c28089a/utils/templates/etc/uwsgi/apps-available/searxng.ini#L69-L76


## nginx / uWSGI-App listens on a Unix socket

https://github.com/searxng/searxng/blob/b3708e4137cf66c7f2966e811c2fae907c28089a/utils/templates/etc/nginx/default.apps-available/searxng.conf:socket#L3


https://github.com/searxng/searxng/blob/b3708e4137cf66c7f2966e811c2fae907c28089a/utils/templates/etc/uwsgi/apps-available/searxng.ini:socket#L69-L73

**Reply by nicfab** [reply]
Thank you.
I think that my settings are correct; I use only searxng.conf because I prefer HHTP instead of socket.
I understand that I can use one of them and not both.
Is it so?

**Reply by return42** [reply]
Its a pair .. when SearXNG listens to upstream on HTTP, proxy has to pass to HTTP .. when SearXNG listens to upstream on a unix-socket, the proxy has to pass to the unix-socket.  Just logical, you just have to set it up correctly in both configurations.

> I changed the installation of my instance from docker to scripts.

The scripts install a uWSGI & nginx setup, where SearXNG listens upstream on a unix-socket.

If you have changed the nginx setup from the scripts, you have also to change the uWSGI setting installed by the scripts .. that was all I wanted to say.

But if it works for you, everything seems OK.

**Reply by nicfab** [reply]
> Its a pair .. when SearXNG listens to upstream on HTTP, proxy has to pass to HTTP .. when SearXNG listens to upstream on a unix-socket, the proxy has to pass to the unix-socket. Just logical, you just have to set it up correctly in both configurations.
> 
> > I changed the installation of my instance from docker to scripts.
> 
> The scripts install a uWSGI & nginx setup, where SearXNG listens upstream on a unix-socket.
> 
> If you have changed the nginx setup from the scripts, you have also to change the uWSGI setting installed by the scripts .. that was all I wanted to say.
> 
> But if it works for you, everything seems OK.

In my case, in `/etc/uwsgi/apps-available/searxng.ini`, I see the following lines:

```
# speak to upstream
# -----------------

# https://uwsgi-docs.readthedocs.io/en/latest/Options.html#plugin-http
# Native HTTP support: https://uwsgi-docs.readthedocs.io/en/latest/HTTP.html

http = 127.0.0.1:8888
buffer-size = 8192

```

Are those lines sufficient, or should I also add the following?

```
socket = ${SEARXNG_UWSGI_SOCKET}
buffer-size = 8192

```

**Comment by return42** [comment]
> In my case, in `/etc/uwsgi/apps-available/searxng.ini`, I see the following lines:

and what is your nginx setup?

> Are those lines sufficient, or should I also add the following?

NO .. linked the templates used by the scripts .. ``${SEARXNG_UWSGI_SOCKET}`` is a place holder for a bash environment when the scripts are running.

**Reply by nicfab** [reply]
> > In my case, in `/etc/uwsgi/apps-available/searxng.ini`, I see the following lines:
> 
> and what is your nginx setup?
> 
> > Are those lines sufficient, or should I also add the following?
> 
> NO .. linked the templates used by the scripts .. `${SEARXNG_UWSGI_SOCKET}` is a place holder for a bash environment when the scripts are running.

In the Nginx file conf I have:

```
server {
  server_name searxng.nicfab.it;
  location / {

    proxy_pass http://127.0.0.1:8888;

    proxy_set_header   Host             $host;
    proxy_set_header   Connection       $http_connection;

    # see flaskfix.py
    proxy_set_header   X-Scheme         $scheme;
    proxy_set_header   X-Script-Name    /searxng;

    # see limiter.py
    proxy_set_header   X-Real-IP        $remote_addr;
    proxy_set_header   X-Forwarded-For  $proxy_add_x_forwarded_for;

    # proxy_buffering  off;
    # proxy_request_buffering off;
    # proxy_buffer_size 8k;

  }

    listen [::]:443 ssl ipv6only=on; # managed by Certbot
    listen 443 ssl; # managed by Certbot
    ssl_certificate /etc/letsencrypt/live/searxng.nicfab.it/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/searxng.nicfab.it/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot

    add_header Strict-Transport-Security "max-age=63072000; includeSubdomains";
    add_header X-Frame-Options DENY;
    add_header X-Content-Type-Options nosniff;
    add_header Onion-Location http://lgmekfnpealdqhnctyg3p5fuelpdwnbqbwnl2ickpdoodzatzgz27ryd.onion$request_uri;


}

server {
    if ($host = searxng.nicfab.it) {
        return 301 https://$host$request_uri;
    } # managed by Certbot


  server_name searxng.nicfab.it;
  #listen [::]:80 ipv6only=off;
  listen 80;
    return 404; # managed by Certbot
}

```

What should I do?

**Reply by return42** [reply]
Nothing, you set `proxy_pass http://127.0.0.1:8888;` and uWSGI `http = 127.0.0.1:8888` .. everything is fine

TBH: we could have saved the whole discussion if you had not claimed:

> I changed the installation of my instance from docker to scripts.

Your nginx setup shown above is not from the [installation scripts](https://docs.searxng.org/admin/installation-scripts.html)!
